core/src/banking_stage/transaction_scheduler/transaction_state.rs

```diff
@@ -0,0 +1,307 @@
+#[allow(clippy::large_enum_variant)]
+pub(crate) enum TransactionState {
+    /// The transaction is available for scheduling.
+    Unprocessed {
+        transaction_ttl: SanitizedTransactionTTL,
+        transaction_priority_details: TransactionPriorityDetails,
+        forwarded: bool,
+    },
+    /// The transaction is currently scheduled or being processed.
+    Pending {
+        transaction_priority_details: TransactionPriorityDetails,
+        forwarded: bool,
+    },
+}
+
+impl TransactionState {
+    /// Creates a new `TransactionState` in the `Unprocessed` state.
+    pub(crate) fn new(
+        transaction_ttl: SanitizedTransactionTTL,
+        transaction_priority_details: TransactionPriorityDetails,
+    ) -> Self {
+        Self::Unprocessed {
+            transaction_ttl,
+            transaction_priority_details,
+            forwarded: false,
+        }
+    }
+
+    /// Returns a reference to the priority details of the transaction.
+    pub(crate) fn transaction_priority_details(&self) -> &TransactionPriorityDetails {
+        match self {
+            Self::Unprocessed {
+                transaction_priority_details,
+                ..
+            } => transaction_priority_details,
+            Self::Pending {
+                transaction_priority_details,
+                ..
+            } => transaction_priority_details,
+        }
+    }
+
+    /// Returns the priority of the transaction.
+    pub(crate) fn priority(&self) -> u64 {
+        self.transaction_priority_details().priority
+    }
+
+    /// Intended to be called when a transaction is scheduled. This method will
+    /// transition the transaction from `Unprocessed` to `Pending` and return the
+    /// `SanitizedTransactionTTL` for processing.
+    ///
+    /// # Panics
+    /// This method will panic if the transaction is already in the `Pending` state,
+    ///   as this is an invalid state transition.
+    pub(crate) fn transition_to_pending(&mut self) -> SanitizedTransactionTTL {
+        match self.take() {
+            TransactionState::Unprocessed {
+                transaction_ttl,
+                transaction_priority_details,
+                forwarded,
+            } => {
+                *self = TransactionState::Pending {
+                    transaction_priority_details,
+                    forwarded,
+                };
+                transaction_ttl
+            }
+            TransactionState::Pending { .. } => {
+                panic!("transaction already pending");
+            }
+        }
+    }
+
+    /// Intended to be called when a transaction is retried. This method will
+    /// transition the transaction from `Pending` to `Unprocessed`.
+    ///
+    /// # Panics
+    /// This method will panic if the transaction is already in the `Unprocessed`
+    ///   state, as this is an invalid state transition.
+    pub(crate) fn transition_to_unprocessed(&mut self, transaction_ttl: SanitizedTransactionTTL) {
```

Comment:
We have to transfer ownership of the `SanitizedTransactionTTL` to the worker threads in order to avoid cloning the data.
So we take ownership when moving it into the `Pending` state, and have to add it back on retry.

Reply:
ah right!